File: firstMissingPositive/firstMissingPositive.java
package leetcode.firstMissingPositive;

class Solution {
    public int firstMissingPositive(int[] nums) {
        buketSort(nums);
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != i + 1) {
                return i + 1;
            }
        }
        return nums.length + 1;
    }

    public void buketSort(int[] nums) {
        int n = nums.length;
        for (int i = 0; i < n; i++) {
            while (nums[i] != i + 1) {
                if (nums[i] <= 0 || nums[i] > n || nums[i] == nums[nums[i] - 1]) {
                    break;
                }
                swap(nums, i, nums[i] - 1);
            }
        }

    }

    public void swap(int[] nums, int i, int j) {
        int tmp;
        tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}

public class MainClass {
    public static int[] stringToIntegerArray(String input) {
        input = input.trim();
        input = input.substring(1, input.length() - 1);
        if (input.length() == 0) {
            return new int[0];
        }

        String[] parts = input.split(",");
        int[] output = new int[parts.length];
        for (int index = 0; index < parts.length; index++) {
            String part = parts[index].trim();
            output[index] = Integer.parseInt(part);
        }
        return output;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        String line;
        while ((line = in.readLine()) != null) {
            int[] nums = stringToIntegerArray(line);

            int ret = new Solution().firstMissingPositive(nums);

            String out = String.valueOf(ret);

            System.out.print(out);
        }
    }
}

File: firstMissingPositive/firstMissingPositive.go
package main

import (
	"fmt"
	"math"
)

func firstMissingPositive(nums []int) int {
	n := len(nums)
	nums = buketSort(nums)
	for i := 0; i < n; i++ {
		if nums[i] != i+1 {
			return i + 1
		}
	}
	return n + 1
}
func buketSort(nums []int) []int {
	n := len(nums)
	for i := 0; i < n; i++ {
		fmt.Println(i)
		fmt.Println(nums)
		for nums[i] != i+1 {
			if nums[i] <= 0 || nums[i] > n || nums[i] == nums[nums[i]-1] {
				break
			}
			nums = swap(nums, i, nums[i]-1)
			fmt.Println(nums)
		}
	}
	return nums
}
func swap(nums []int, i, j int) []int {
	tmp := nums[i]
	nums[i] = nums[j]
	nums[j] = tmp
	return nums
}

func main() {
	nums := []int{3, 4, -1, 1}
	// nums := []int{-5, 1000}
	// fmt.Println(nums[10])
	fmt.Println(firstMissingPositive(nums))
	fmt.Println(math.MaxInt32)
}
